### DISCUSSION TITLE
UI idea: exploiting affixation

### DISCUSSION THREAD

**Discussion by bdarcus** [emacs-citar/citar#28] [Ideas]
I started this before I implemented the idea. I've now merged it, and should be a good base to iterate on.

### Background

I thought I'd write down some ideas I've thought about as I've put this code together, and in the process learned about bibtex-completion, completing-read, etc., and started a PR.

I've created this issue to track it: https://github.com/bdarcus/bibtex-actions/issues/33.

Right now, the UI is very similar to helm-bibtex and ivy-bibtex. It presents a single, undifferentiated, formatted propertized display string, but internally searches using a different string.

This works well.

But there's room with completing-read to use annotations for secondary bits of supporting information. Affixation (recently added to Emacs 28) seems the forward-looking way to do this (see: https://github.com/minad/marginalia/issues/50#issuecomment-798896705).

**Is there value in rethinking the UI a bit to take advantage of that, and adding an affixation function to support it, either here or in a separate package?**

If yes, how would that look, and what should that function generate?

My suggestions:

### Prefix

Perhaps prefixes with [nice icons or icon fonts](https://github.com/rougier/emacs-svg-icon) (see [also](https://github.com/domtronn/all-the-icons.el#usage)) to indicate presence of pdfs, notes, or links, like this?

![Screenshot from 2021-03-13 15-04-00](https://user-images.githubusercontent.com/1134/111042869-765a4200-840d-11eb-8911-7ea89de7f47b.png)

The icons would need to be aligned, however, so they at least appeared as in columns.

### Suffix

Maybe a suffix with the key, type, and maybe tags?

If yes, the actually displayed string would be more compact (say just author, date, title?) to make room.

Something like this:

![Screenshot from 2021-03-14 08-11-11](https://user-images.githubusercontent.com/1134/111068064-7c97ff00-849d-11eb-8449-0030db0623a6.png)

Maybe it's user-configurable?

Also, early in these discussions, marginalia (and consult) author @minad [suggested the possibility of a separate package for annotation](https://github.com/tmalsburg/helm-bibtex/issues/353#issuecomment-766832311). Does that still make sense?

~If someone does want to experiment with this, do post screenshots.~ This is now fully implemented, with configurable main display and suffix.

See also #20.

**Comment by apc** [comment]
So I think this is relevant to this, though I’m not quite sure. But here it goes anyway. Is there a way to see on the list of entries which of them have associated notes and/or PDFs *without* using emacs 28? Something like what `ivy-bibtex` offers? I’m very reluctant to give up on using the Mac Ports build, which only offers Emacs 27 at the moment AFAIK.

**Reply by bdarcus** [reply]
Yeah, you can use the same template rules as you do with bibtex-completion, for the same experience. Just add those fields to your suffix template.

**Reply by apc** [reply]
So essentially set the value of `bibtex-actions-template-suffix` to that of `bibtex-completion-display-formats`?

**Reply by bdarcus** [reply]
That, and/or the main template, depending on your preference of course.

If you wanted it more visually prominent, you could put those fields at the beginning or end of the main template.

If more subtle (depending on theme, or if and how you customize the suffix face), put in the suffix.

BTW, I'm adding a test script derived from the selectrum ones in #90; will update the README there.

**Reply by apc** [reply]
OK, I'll have a look at some point and let you know if I am able to help with the `use-package` stuff.

**Reply by bdarcus** [reply]
Is it possible for you to move this to a new thread, for easier later discovery?

Guessing not, since I can't.

In the meantime, what does your main template look like?

**Reply by bdarcus** [reply]
> I can delete the comment and start a new one if you want me to.

Yes, please. People won't be able to find this otherwise.

In the meantime, I put in a request to github to allow me to promote comments to new threads.

To be clear, you are on Emacs 27?? I don't see how you can have that first symbol if you do.

**Reply by apc** [reply]
I am! 27.2... ¯\_(ツ)_/¯
